### PR TITLE
Some improvement to the avalonia vulkan backend

### PR DESCRIPTION
**Pull request by emmauss**
1. Fix up crash that occurs when you toggle fullscreen
2. Revert vsync changes. Nvidia optimus gpu's become unstable when present mode is changed from mailbox to immediate mode on the same window.

### CODE REVIEW

File: Ryujinx.Ava/Ui/Controls/VulkanRendererControl.cs [hunk 1]
```diff
@@ -34,7 +35,21 @@ public override void DestroyBackgroundContext()
 
         protected override ICustomDrawOperation CreateDrawOperation()
         {
-            return new VulkanDrawOperation(this);
+            var newDrawOperation = new VulkanDrawOperation(this);
+
+            if(_currentDrawOperation != newDrawOperation)
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Controls/VulkanRendererControl.cs:40]:
```suggestion
            if (_currentDrawOperation != newDrawOperation)
```

File: Ryujinx.Ava/Ui/Controls/VulkanRendererControl.cs [hunk 2]
```diff
@@ -73,7 +90,18 @@ public VulkanDrawOperation(VulkanRendererControl control)
 
             public void Dispose()
             {
+            }
+
+            public void Destroy()
+            {
+                if(_isDestroyed)
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Controls/VulkanRendererControl.cs:97]:
```suggestion
                if (_isDestroyed)
```

File: Ryujinx.Ava/Ui/ViewModels/MainWindowViewModel.cs [hunk 3]
```diff
@@ -929,6 +930,13 @@ public static void OpenLogsFolder()
 
         public void ToggleFullscreen()
         {
+            if(Environment.TickCount64 - _lastFullscreenToggle < 500)
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/ViewModels/MainWindowViewModel.cs:933]:
```suggestion
            if (Environment.TickCount64 - _lastFullscreenToggle < 500)
```

Comment (gdkchan) [reply Ryujinx.Ava/Ui/ViewModels/MainWindowViewModel.cs:933]: Why was this check added btw?

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs [hunk 4]
```diff
@@ -236,35 +302,65 @@ public override unsafe void Present(ITexture texture, ImageCrop crop, Action<obj
...
+                    _states[_nextImage],
+                    cbs.GetFence());
+
+                _presentedImages[_nextImage] = info;
+            }
+            else
+            {
+                info.Fence = cbs.GetFence();
+            }
+
```
Comment (gdkchan) [inline Ryujinx.Graphics.Vulkan/ImageWindow.cs:355]: Double empty line?

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs [hunk 5]
```diff
@@ -300,7 +396,8 @@ public override unsafe void Present(ITexture texture, ImageCrop crop, Action<obj
                 barrier);
         }
 
-        private void CaptureFrame(TextureView texture, int x, int y, int width, int height, bool isBgra, bool flipX, bool flipY)
+        private void CaptureFrame(TextureView texture, int x, int y, int width, int height, bool isBgra, bool flipX,
```
Comment (gdkchan) [inline Ryujinx.Graphics.Vulkan/ImageWindow.cs:399]: Why was this changed?

File: Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDisplay.cs [hunk 6]
```diff
@@ -23,8 +22,17 @@ internal class VulkanDisplay : IDisposable
         private Image[] _swapchainImages;
         private VulkanDevice _device { get; }
         private ImageView[] _swapchainImageViews = new ImageView[0];
-        private bool _vsyncStateChanged;
-        private bool _vsyncEnabled;
+        private bool _surfaceChanged;
+
+        public bool NotifySurfaceChanged {
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDisplay.cs:27]: `{` on new line.

File: Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDisplay.cs [hunk 7]
```diff
@@ -23,8 +22,17 @@ internal class VulkanDisplay : IDisposable
         private Image[] _swapchainImages;
         private VulkanDevice _device { get; }
         private ImageView[] _swapchainImageViews = new ImageView[0];
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDisplay.cs:24]:
```suggestion
        private ImageView[] _swapchainImageViews = Array.Empty<ImageView>();
```

File: Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDevice.cs [hunk 8]
```diff
@@ -12,12 +12,12 @@ public VulkanDevice(Device apiHandle, VulkanPhysicalDevice physicalDevice, Vk ap
             InternalHandle = apiHandle;
             Api = api;
 
-            api.GetDeviceQueue(apiHandle, physicalDevice.QueueFamilyIndex, 0, out var queue);
+            // pick last queue in family
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDevice.cs:15]:
```suggestion
            // Pick last queue in family.
```

File: Ryujinx.Ava/Ui/Backend/Vulkan/Skia/VulkanRenderTarget.cs [hunk 9]
```diff
@@ -8,8 +9,14 @@ namespace Ryujinx.Ava.Ui.Backend.Vulkan
 {
     internal class VulkanRenderTarget : ISkiaGpuRenderTarget
     {
-        public GRContext GrContext { get; set; }
-        
+        public GRContext GrContext {
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Backend/Vulkan/Skia/VulkanRenderTarget.cs:12]: `{` on new line.

File: Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDevice.cs [hunk 10]
```diff
@@ -12,12 +12,12 @@ public VulkanDevice(Device apiHandle, VulkanPhysicalDevice physicalDevice, Vk ap
...
+            api.GetDeviceQueue(apiHandle, physicalDevice.QueueFamilyIndex, Math.Min(1, physicalDevice.QueueCount - 1), out var queue);
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDevice.cs:16]: Why the `Min`? It doesn't looks like this actually picks the last one always.

File: Ryujinx.Ava/Ui/ViewModels/MainWindowViewModel.cs (hunk 3 above)
Comment (emmauss) [reply Ryujinx.Ava/Ui/ViewModels/MainWindowViewModel.cs:933]: Hotkeys aren't rate limited, so holding the fullscreen key will cause the window to rapidly switch between modes, which will cause the swapchain and images to be recreated at a high rate. It was one of the causes for the fullscreen crash, even after ensuring the game renderer resources are properly recreated when presentation begins

File: Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDevice.cs (hunk 10 above)
Comment (emmauss) [reply Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDevice.cs:16]: in hindsight, this doesn't do much since queues aren't shared between devices.

File: Ryujinx.Ava/Ui/ViewModels/MainWindowViewModel.cs (hunk 3 above)
Comment (gdkchan) [reply Ryujinx.Ava/Ui/ViewModels/MainWindowViewModel.cs:933]: Was this addressed?

Comment (emmauss) [reply Ryujinx.Ava/Ui/ViewModels/MainWindowViewModel.cs]: It's done now.

File: Ryujinx.Ava/Ui/Controls/RendererControl.cs [hunk 11]
```diff
@@ -94,7 +84,7 @@ protected void OnRendererInitialized()
         }
 
         public void QueueRender()
-        {
+        {
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Controls/RendererControl.cs:87]:
```suggestion
        {
```

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs [hunk 12]
```diff
@@ -343,23 +405,89 @@ public override void Dispose()
...
+        public Device Device { get; }
+        public PhysicalDevice PhysicalDevice { get; }
+        public ulong MemorySize { get; }
+        public ulong MemoryOffset { get; }
         public Semaphore ReadySemaphore { get; }
         public Semaphore AvailableSemaphore { get; }
+        public Extent2D Extent { get; }
+        public ImageState State { get; internal set; }
+        internal PresentImageInfo(
+                Auto<DisposableImage> image,
```
Comment (gdkchan) [inline Ryujinx.Graphics.Vulkan/ImageWindow.cs:460]: Indentation is off here.

**Review by gdkchan — COMMENTED** [review]
Will try to review and test more thoroughly after Vulkan review is complete.

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs (hunk 12 above)
Comment (emmauss) [reply Ryujinx.Graphics.Vulkan/ImageWindow.cs]: Addressed

**Comment by gdkchan** [comment]
Since it was already merged, please open a PR on the main repo.
